core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -82,7 +84,7 @@ trait Endpoint[A] { self =>
 
       final def apply(input: Input): Endpoint.Result[B] = self(input) match {
         case EndpointResult.Matched(rem, trc, out) =>
-          EndpointResult.Matched(rem, trc, out.flatMapF(this))
+            EndpointResult.Matched(rem, trc, out.map(o => o.traverse(fn)).flatMap(f => Task.async(f)))
```

Comment:
I wonder if we can keep `flatMap(this)` in place for the sake of saving allocations.

If we modify `apply` as follows, maybe it should work:

```scala
final def apply(oa: Output[A]): Task[Output[B]] = Task.async(oa.traverse(fn))
... 

// and then 
out.flatMap(this)
```

Reply:
I'll give it a go later today

Reply:
I think I found a solution closest to your suggestion, couldn't edit the `final def apply(oa: Output[A]): Task[Output[B]]` because it should remain returning a `Future[_]` to not break the interface (or am I missing something)

Reply:
You can change the interface - it's only used to reduce closure allocations. It's a trick I learned from Flavio (Arrows author) and he called it "merge interfaces". Here is an example:

```scala
class Foo extends AtomicInteger with (Try[Int] => Unit) {
  def apply(ti: Try[Int]): Unit = ti match {
    case Return(i) => set(i) // AtomicInteger.set
    case Throw(_) => 
  }
}

val i: Future[Int] = ???
i.respond(atomicInteger)
```

So instead of allocating an extra closure for `i.respond { case Return(_) => }` and updating a given atomic integer we just use a single object (a smash of both, a closure and an atomic integer) as both things. This helps to save quite some allocations.

Reply:
Thanks for the pointer, it worked